### ISSUE TITLE
More ergonomic API to avoid flattening/unflattening of high-level variables

### ISSUE THREAD

**Issue by oleganza** [dalek-cryptography/bulletproofs#218]
**Problem**

If you have structured high-level variables (e.g. lists of tuples, like inputs and outputs of confidential values for a shuffle gadget), then it's pretty annoying to flatten them in one list of scalars, which is turned into commitments and variables, and then variables are used to construct back the structured set of "Allocated Values" together with putting back their assignments from the original structured input.

**Proposal**

Instead of passing a list of all scalars and a list of their blinding factors, allow individual commitment of them:

```
prover.commit_external_variable(v,v_blinding) -> (RistrettoPoint, Variable)
``` 

which internally creates a commitment, writes it to the transcript, allocates variable and returns the commitment and a variable.

Then, the user can traverse their structured set of secrets and commit them one by one, creating allocated+assigned types (specific to their protocol) along the way. 

E.g. if the user has a list `Vec<Value>`, it can be `map`'d to `Vec<AllocatedValue>` much more easily than what we have to do today.

**Comment by cathieyun** [comment]
This sounds helpful!

**Comment by hdevalence** [comment]
Hmm, right now the external commitments are passed into the constraint system constructor in order to ensure that it's not possible to start building constraints without having declared all of the inputs to the constraint system.  Wouldn't this function break that, unless it was on a `Builder` of some kind?

**Comment by cathieyun** [comment]
I see the concern - if this were implemented there would have to be two "states", one where the user is still adding high-level variables, and one where the user is done and wants to allocate / use low-level variables. And there would have to be some kind of guarantee that the two states stay separated.

**Comment by oleganza** [comment]
Probably no need for separate Builder type: have `prove` method consume self.

**Comment by hdevalence** [comment]
The `prove` method does consume `self` already, but that doesn't deal with this issue, which is that the high-level variables (constraint system inputs) need to all be set before the constraints are added (which has to happen before proving).

**Comment by oleganza** [comment]
Oops, you are right. Then we can have `Prover` as a starting type which is then converted to `ProvingCS` by the `prove` method that takes a closure with CS as an argument, consumes the prover and returns the proof (the Prover being the “builder”). Closure would let us avoid having two steps “convert to CS” and “make proof”.

**Comment by oleganza** [comment]
Fixed in #221.
